subsys/net/lib/app/net_app.c

```diff
@@ -506,13 +516,104 @@ struct net_context *select_client_ctx(struct net_app_ctx *ctx,
 #endif /* CONFIG_NET_APP_CLIENT */
 
 #if defined(CONFIG_NET_APP_SERVER)
+static struct net_context *get_server_ctx(struct net_app_ctx *ctx,
+					  const struct sockaddr *dst)
+{
+	int i;
+
+	for (i = 0; i < CONFIG_NET_APP_SERVER_NUM_CONN; i++) {
+		struct net_context *tmp;
+		u16_t port, rport;
+
+		if (!ctx->server.net_ctxs[i] ||
+		    !net_context_is_used(ctx->server.net_ctxs[i])) {
+			continue;
+		}
+
+		tmp = ctx->server.net_ctxs[i];
+
+		if (!dst) {
+			if (tmp->net_app == ctx) {
+				NET_DBG("Selecting net_ctx %p for NULL dst",
+					tmp);
+				return tmp;
+			}
+
+			continue;
+		}
+
+		/* Serve IPv6 first if the user does not care */
+		if (IS_ENABLED(CONFIG_NET_IPV6) &&
+		    (dst->sa_family == AF_UNSPEC ||
+		     (tmp->remote.sa_family == AF_INET6 &&
+		      dst->sa_family == AF_INET6))) {
+			struct in6_addr *addr6 = &net_sin6(dst)->sin6_addr;
+			struct in6_addr *remote6;
+
+			remote6 = &net_sin6(&tmp->remote)->sin6_addr;
+			rport = net_sin6(&tmp->remote)->sin6_port;
+			port = net_sin6(dst)->sin6_port;
+
+			if (dst->sa_family == AF_UNSPEC) {
+				if (tmp->net_app == ctx) {
+					NET_DBG("Selecting net_ctx %p for "
+						"AF_UNSPEC port %d", tmp,
+						ntohs(rport));
+					return tmp;
+				}
+
+				continue;
+			}
+
+			if (net_ipv6_addr_cmp(addr6, remote6) &&
+			    port == rport) {
+				NET_DBG("Selecting net_ctx %p for AF_INET6 "
+					"port %d", tmp, ntohs(rport));
+				return tmp;
+			}
+		}
+
+		if (IS_ENABLED(CONFIG_NET_IPV4) &&
+		    tmp->remote.sa_family == AF_INET &&
+		    dst->sa_family == AF_INET) {
+			struct in_addr *addr4 = &net_sin(dst)->sin_addr;
+			struct in_addr *remote4;
+
+			remote4 = &net_sin(&tmp->remote)->sin_addr;
+			rport = net_sin(&tmp->remote)->sin_port;
+			port = net_sin(dst)->sin_port;
+
+			if (dst->sa_family == AF_UNSPEC) {
```

Comment:
Without (dst->sa_family == AF_UNSPEC) in the prior check this will never be true.